feat: Add support for multiple services in aws-proxy

liustve opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
- Current AWS proxy design is limited to a single AWS service because it uses a single ServiceName and AWSEndpoint configuration that applies to all incoming requests. This change adds an optional `AdditionalRoutingRules` configuration that allows routing specific APIs to different services which override the now "default" region and endpoint settings. The AWSEndpoint field in routing rules is optional and will auto-resolve from the service name and region when not specified, with unspecified fields falling back to the top-level proxy configuration.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
None

<!--Describe what testing was performed and which tests were added.-->
#### Testing
- Unit tests added
- Manual E2E tests

Example `config.yaml`:
```
extensions:
  awsproxy:
    endpoint: 0.0.0.0:2000
    service_name: "xray"
    region: "us-west-2"
    additional_routing_rules:
      - paths: ["PutLogEvents", "CreateLogGroup", "DescribeLogStreams"]
        service_name: "logs"
        region: "us-east-1"
        role_arn: ""
      - paths: ["PutMetricData"]
        service_name: "monitoring"
        region: "us-west-2"
      - paths: ["slos"]
        service_name: "application-signals"
        region: "us-east-1"
        aws_endpoint: "https://application-signals.us-east-1.api.aws"

receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318

processors:
  batch:

exporters:
  debug:
    verbosity: detailed

service:
  extensions: [awsproxy]
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug]
    logs:
      receivers: [otlp]
      processors: [batch]
      exporters: [debug]
```

X-Ray:
```
curl -X POST http://localhost:2000/GetSamplingRules
{"NextToken":null,"SamplingRuleRecords":[{"CreatedAt":0.0,"ModifiedAt":0.0,"SamplingRule":{"Attributes":{},"FixedRate":0.05,"HTTPMethod":"*","Host":"*","Priority":10000,"ReservoirSize":1,"ResourceARN":"*","RuleARN":"arn:aws:xray:us-west-2:571600841604:sampling-rule/Default","RuleName":"Default","ServiceName":"*","ServiceType":"*","URLPath":"*","Version":1}}]}
```

Logs:
```
curl -X POST http://localhost:2000/DescribeLogStreams \
  -H "Content-Type: application/x-amz-json-1.1" \
  -H "X-Amz-Target: Logs_20140328.DescribeLogStreams" \
  -d '{"logGroupName":"/aws/apigateway/welcome","limit":2}'
{"logStreams":[{"arn":"arn:aws:logs:us-east-1:571600841604:log-group:/aws/apigateway/welcome:log-stream:2025-10-23_20-11-00","creationTime":1761250260335,"firstEventTimestamp":1761250260255,"lastEventTimestamp":1761250260255,"lastIngestionTime":1761250260348,"logStreamName":"2025-10-23_20-11-00","storedBytes":0,"uploadSequenceToken":"49039859648811597105500463748228744792432845772976541179"}]}
```
Application Signals:

```
curl -X POST "http://localhost:2000/slos?MaxResults=1" -d '{}'
{"NextToken":"AAMA-EFRSURBSGdRWUNwRVNaQ2FKNXlhTW5HdkoxZEZOcCtscUFXbnREVXFDZGxFU25rc0dBSGZKNEFjSjg1UzB4RysvSHM0c1ZIM0FBQUFmakI4QmdrcWhraUc5dzBCQndhZ2J6QnRBZ0VBTUdnR0NTcUdTSWIzRFFFSEFUQWVCZ2xnaGtnQlpRTUVBUzR3RVFRTWJ1WmhXSStXSjRyU0ZrZGpBZ0VRZ0R2WVhNRFRvUEczb1ZyclVKQkJXTUppTzQveVhSY040dmQ0cUFMWnhmZ0hWWGxKamRQeEJld21yZHBNZ2RMWXgrUzZNVEJpUklwYlc1ZEczQT09nOv3uSIDV2AQ2mFGJ5z2xB3ktB40MgqoNe3cJNICDOPuvYjKhsKCRo6Zrto_fV_67JVEdPxVVjI1Cwj8X0rBUBSgQDlUF1x3slxRAmOlIRAa6KIuAMJFNoCmIy6I3FuKRT5UAtDveJN97VIhc5mFbZy-FN2Q8Q3fWeJDC62ThK2lxR79OOdEB1MnPLNfOuyt2V1ILPmS8S-zX68s50FxXqwdCPSEOzGJaqvyA5c1ywRkkPEBPl_jwJmeAtElcg5mOh7_uS45bHC0lQPBqZyJJcgWDHTRrqE6hA1OVxZhliAAd13Aag4Y-XsAfDfViv1h4jzdH7Vo78SAbakwzHxGXgBk1zPaSiqWzTWkE0o6umkRazE1Oa3MKhKHXSuspITjuWMeUnj213WNOaP8eQnqQIoibICRAVstTFChGuNTfvE1ij6fXyPjH6ehIeIs59W6pi6KgNiteT3ZdohdzovDZPeCcmde4kJATyttIHNF-QNyuNPV5kB3ubD_3frsxCcug6Z0Eh6qmlTMas-1kpsIVyEUTOi3qZsH4AG-2xulXtWBBoLJvSNzyQWGLBbhpvrLmWcZckGRLVSXfTzPuneO-uQWqDzBc5jehbvIu2jk1QYWmkwHh4tuaTyXQZ22NFAvj8CF19MHV2d04ECTzTZEG238WW97mCTEcZaf387NqAf0JwKXYxNZQRkMRIfSvsPx0TwAN_jTctdaZz1ob5zI4nX8bWYSTVp8XuaBBkTrOTsFm6Z4a0tmhi2ucJ8ODRhtsDfu_s_1rU4hyNfUO7qS7YP9fx9Z2kLKJLf8U4lRBUEO7TfDaVOhUd6VxOA7CUkVZnViGrpWoHag3eAOATxTfmaDr_qFMoZTU-nMAt5iQRNQGmW5Olie3S2kkCfmb18mZEH_jl_x2GDDmT8uP1id7rAEh9WSFEDvpt210mAMbNNiXLu1ks6AVhAhyUmdOG9fzg4SZtOBQl7qkEOCzIJNVHRZoCo4Wg-MVHL1Lik2nfQIYuG7DdUdMKM9JgPY5m8QsMiQz6wWARbpEdu0uZmPlvaJaVScmdzUdOl-hQ4oUL3ohAutBsgyiOfX_eLrb3ZcAkwiYjOpeHwsg9k2e5FQkCzqUNpJH4W9mlEckimOdpwZGwOCyJFfamhe2nK7MAztoReDGZMmMZxtIc8EOcAJ-0jT15dAPDspuqdyezu2ERfLT0tQbhJ66gZgJDpdm6HECexQmz455Wh0OO8xZoG7RUegTmiNP-SEDRHsBXUEPAlkQnSR6bJUN9w8q6CKTZ7Jb_O6HhAHV9RJnzPCQ1lnS63h7gdHT8HMSiaM9DYcfdA_gBpfor0OFvPnkWLz6HUdBn2VzuvamWvOZ9E_jlwy1X1O9WF35-irxJigFG9c94vzpfgV8cWc6GXN9olwafOtJ8UoUmAh0mUlfL1n-08poZpxNLAq2hAdT61cChbdXa4uYMU0fpu5LMA_rnWfU7aOLao0aV0ANd7ojhjDkbXPymO9NCSd78O5Q9PsiwrURsGagsMH0D1mp2oUGTyO4ConIopkD20zxY4Zh2SZ9L0Wn4wJVrphKUakIKl0miyIv-CLh2mULXj0O8wN6AQpAycz1sS8tRKIytZNBQ0obafPt-Y9H1lm3Pvb9unXlt5mgn7x6Uf1g6_HOxoT9oO0dKWoh0530JgM3ltVkygJJGdSFkuGGdpoIxF5Gmd_1VBRI55vzPUSgOwXNV6AHDenpufLC_5zIEdJItuoUIWN7OEViimQetSdaifEGgeRI1Onh6DepGA8K3oFOpE=","SloSummaries":[{"Arn":"arn:aws:application-signals:us-east-1:571600841604:slo/TestNutritionAgentSLO3","CreatedTime":1.761768148904E9,"DependencyConfig":null,"EvaluationType":"RequestBased","KeyAttributes":null,"MetricSource":null,"MetricSourceType":"CloudWatchMetric","Name":"TestNutritionAgentSLO3","OperationName":null}]}
```


-------------

Tested with `CWAgent`:

```
./build/bin/linux_amd64/amazon-cloudwatch-agent   --config empty.toml   --otelconfig /home/ec2-user/opentelemetry-collector-contrib/config.yaml
2025-12-22T21:37:37Z W! Failed to load environment variables due to cannot read env config file env-config.json due to: open env-config.json: no such file or directory
2025-12-22T21:37:37Z I! Starting AmazonCloudWatchAgent CWAgent/Unknown (go1.25.5; linux; amd64) with log file  with log target file
2025-12-22T21:37:37Z I! AWS SDK log level not set
2025-12-22T21:37:37Z I! {"caller":"service@v0.124.0/service.go:199","msg":"Setting up own telemetry..."}
2025-12-22T21:37:37Z I! {"caller":"builders/builders.go:26","msg":"Development component. May change in the future."}
2025-12-22T21:37:37Z I! {"caller":"service@v0.124.0/service.go:244","msg":"Skipped telemetry setup."}
2025-12-22T21:37:37Z I! {"caller":"service@v0.124.0/service.go:266","msg":"Starting CWAgent...","Version":"Unknown","NumCPU":16}
2025-12-22T21:37:37Z I! {"caller":"extensions/extensions.go:41","msg":"Starting extensions..."}
2025-12-22T21:37:37Z I! {"caller":"extensions/extensions.go:45","msg":"Extension is starting..."}
2025-12-22T21:37:37Z I! {"caller":"awsproxy@v0.124.1/extension.go:39","msg":"X-Ray proxy server started on 0.0.0.0:2000"}
2025-12-22T21:37:37Z I! {"caller":"extensions/extensions.go:62","msg":"Extension started."}
2025-12-22T21:37:37Z I! {"caller":"otlpreceiver@v0.124.0/otlp.go:116","msg":"Starting GRPC server","endpoint":"0.0.0.0:14317"}
2025-12-22T21:37:37Z I! {"caller":"service@v0.124.0/service.go:289","msg":"Everything is ready. Begin running and processing data."}
```

```
$ curl -X POST "http://localhost:2000/slos?MaxResults=1" -d '{}' -v 2>&1
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:2000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:2000...
* Connected to localhost (::1) port 2000
> POST /slos?MaxResults=1 HTTP/1.1
> Host: localhost:2000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Length: 2
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 200 OK
< Content-Length: 2460
< Content-Type: application/json
< Date: Mon, 22 Dec 2025 21:38:07 GMT
< X-Amzn-Requestid: 2c8cc8ab-336e-490d-9297-0eb3fe5153a9
< 
{"NextToken":"AAMA-EFRSURBSGdRWUNwRVNaQ2FKNXlhTW5HdkoxZEZOcCtscUFXbnREVXFDZGxFU25rc0dBR3RFVWpua2FaTWRYaWtYTVZrRG1aZ0FBQUFmakI4QmdrcWhraUc5dzBCQndhZ2J6QnRBZ0VBTUdnR0NTcUdTSWIzRFFFSEFUQWVCZ2xnaGtnQlpRTUVBUzR3RVFRTTlWdkdxUHVOcXkyenV0NG1BZ0VRZ0RzZ0l1dHZ5dkd5VTlBbE9NKzZUWkxVSVVKSldiMDllOERDdXp5Mmx2d0JzeEF0MktpN1JNZFAwS3Jxb25VWnpNcEgzVlZWeE4yYk93dDVsZz09llbqQ4am3syt6nBSzAhManEFBhXWpfQQXKEMaX-YcKHRefMSU-0nzbrBIhvkKF8cfJbtNlPjFmrBqZqKi_Dhiq0036yBm3juJrsSc9uA6KUI1RR72sT2IFX2MwBITO8GNQmWqeDWZFS70k_kG-IHDZKFq5Osc8182q6iqAfQp6g71vR-IPrPIU79LOGEawCKN7FPo3T4jgH8MPYpVYMiZjb4kizA3KTi0f9xE2pITvReFgEqWUfkhv4uv0jt7QYbtVNIPdaipYI0-_o99XyiueUEv2GhjZUnCTx-iqyAFN3pahQUdfmyySe4gZTFZfIvbyCowABEjqjhkxc4q-4HhfIkQA0ZPIFmpGP_P4M3c-vQTXVGueNFGl3jXZX814pOOTImhLeO5zy_G2RqrA5JDbgvFIRqWuuipYwRzLv1b_McRn-1RJ4EdmhUZy9fWWR50aupj0Y3VL5PK8d-XGwpHFr_bFhudETpfA6q9G5qAxHya75Cosvu0be0IfpKSkHCTPgMtkpn1ztLJoduZ8GExH-g0V93BELU9OoU1RoNoBCIg_gzr-sziwyVDfltveKf8BKqDdwxesFZtxEWdyd-D-bGqWmNcWbsVJ5dWMnDRTxSA5uKdK6ZA2UAkdkawLkhptxRmm6Fepk0Z6szeFOdMWxuQ_RZCIX1ULWdL1RuSIMTiXbvpOmABhO_nSLIc_b-MwfMk8GzAMoe9_xZnj3rRYNb1bI6-5_NZ2iTUnd7Hp0KtCc5zknacujdU91FtCO14huuXfb0USwS69FFamaYy8UauNUYWBSMDyI-_vVwe2c2A1fKZxRXziDCNl6HWfdRR8A9uekTY7-OLzxx1nW-HdRS-qWV0_VLIRkDrw2Y_48wH9WkHaY-7iN-ZyTepFYMHaQnRB3-13wkX8wcIn4g2Q-YAXyojQVH1FA2IgQYnXfhxMV-3UdmLt48a_FL8C7JTmedqjrs9yXC2MMVPB5DFAxnRo6NNwbhvTWHrgWPpE0RSFK6QRDDaoPUGjuz9iGw0y57ZcDqVCQTzgbDUC-jxpD1y1OMCn-ZrshbBiOKheGeJHt38CIwULrmKKUwgRUM6I-QrjK1aBMgGCm9XwbsmmsxTvVka83DiJdVkCdvwLF0GAfId_Sd1K0eQSaRfhvobzmOcrylg2oz8jFkOB0D5L5oJGwCvXPcMS7vqPs9XtXrdoOcXkacQpHexWIPBFqOZ9c88d_rWVrwo00UYSalz7nJnB1van9ANWota6-Xp8A9NZntedOEGhsWvDMDQd1iR4xZqEEeXIvgLDw2MJahyVC0XXLUU3Lf9o-tkr5PclB-JuZvQ4mzCevl6_w9Ouh_flB1FapxtBFbnXF2Y8aA_X8Fred9Xe5rwmIpy1iqh6vzdaoGCRKPKndt27Gi-HB9e3wN85PN-Nn2n1xHpCAos7Lc2oHZItUaGlF3XJyLLQiyung_42E1XxZ-cz3jNN6pm4A7ac4oOSSaoBGhTcX6J41djnxwxjPFmeLc-JcB43XKWBANGQboTbVIyDApV9hS1TqhyH1JU7Vy0Hn8Hz7hQblDRVzHmbcrXRYJEYyBJtEmSECHFBGqwUOqqW6nJ9DV_LYkqWnszZ3S72teg7DvQTkOHbvTU513tbwVyi9UwHOILdIbr8lnpvpsqcPljBhMA1Lbk_XOr8ahEyA7jUaTMyXQGud1-IeQpf* Connection #0 to host localhost left intact
6qKvcw8EylUpAJXfEulCvvnEsdrPexX3wwybdsies":[{"Arn":"arn:aws:application-signtedTime":1.761768148904E9,"DependencyCnull,"MetricSource":null,"MetricSource
amazon-cloudwatch-agent]$ 
```

<!--Describe the documentation added.-->
Added documentation for new configuration in `aws-proxy`


<!--Please delete paragraphs that you did not use before submitting.-->

liustve committed:
add support for multi service proxy

liustve committed:
Merge branch 'aws-cwa-dev' into aws-proxy-multi-server-support

liustve committed:
refactor tests

liustve committed:
lint fix + naming

vastin commented on internal/aws/proxy/cfg.go:
```diff
@@ -62,6 +62,30 @@ type Config struct {
+type ServiceConfig struct {
+	// APIs is a list of API names to match against the request.
+	// Example: "slos", "GetSamplingRules", "DescribeLogStreams"
+	APIs []string `mapstructure:"path"`
```
The "path" is inconsistent with "apis" in README.md file.

liustve replied:
Good catch, I renamed it to paths. I think it makes more sense this way given that the way to call `ListSlos` isn't directly through the API name but something like:
```
curl -X POST http://localhost:2000/slos \
  -H "Content-Type: application/json" \
  -d '{"MaxResults": 5}'
```

vastin commented on internal/aws/proxy/cfg.go:
```diff
@@ -62,6 +62,30 @@ type Config struct {
+	Region string `mapstructure:"region"`
+
+	// RoleARN overrides the default IAM role for this service.
+	RoleARN string `mapstructure:"role_arn"`
```
The role arn is defined but never used.

liustve replied:
This was a miss, I added code to support this.

vastin commented on internal/aws/proxy/server.go:
```diff
@@ -90,10 +96,47 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
-			req.URL.Host = awsURL.Host
-			req.Host = awsURL.Host
+			apiName := req.URL.Path
+			// strip the "/" from the request path
```
nit: use `strings.TrimPrefix()` instead.

liustve replied:
done

vastin commented on internal/aws/proxy/server.go:
```diff
@@ -161,3 +204,21 @@ func setResolverConfig() func(*endpoints.Options) {
+			// Technically duplicate paths shouldn't happen, but if the same path is configured
+			// for multiple services, the first service wins.
+			if _, exists := apiMap[path]; !exists {
+				apiMap[path] = &route
```
Why is the leading '/' not stripped for `path` here?

liustve replied:
Added an explanation, but in short this is parsed from the config file which will not have the leading /.
When parsing it from the URL on the other hand it will have the "/"

liustve committed:
add support for role arn override

liustve committed:
rephrase readme

mxiamxia commented on internal/aws/proxy/cfg.go:
```diff
@@ -62,6 +62,30 @@ type Config struct {
+	// AdditionalRoutingRules defines optional routing rules for multi-service support.
+	// Each route can override service name, region, role, and endpoint.
+	// If a field is not provided, defaults to the top level configuration.
+	AdditionalRoutingRules []ServiceConfig `mapstructure:"additional_routing_rules,omitempty"`
```
`ServiceConfig` naming seems very generic, is it better just using `RoutingRule` or `AdditionalRoutingRule`?

liustve replied:
Good point I renamed it to `RoutingRule`

mxiamxia commented on extension/awsproxy/README.md:
```diff
@@ -67,3 +72,13 @@ The AWS service endpoint which this proxy forwards requests to. If not set, will
+
+### additional_routing_rules (Optional)
+A list of routing rules to forward requests to different AWS services based on the operation. Each rule can override the service name, region, role ARN, and endpoint if specified, other wise it will default to the top level config.
+
```
Will additional rule override the default top level proxy config or it will be set as additional rule in addition to the top level config? I think it should be in addition instead of overriding as you described

liustve replied:
Good point I think

"Additional" would imply these rules are added to the default behavior. The default top level routing still works and these rules extend it for other services

"Override" would imply replacing/overwriting the default behavior entirely.

Changed.

mxiamxia commented on internal/aws/proxy/server.go:
```diff
@@ -90,10 +99,46 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
+				} else {
+					resolved, err := getServiceEndpoint(&aws.Config{Region: &region}, serviceName)
+					if err != nil {
+						logger.Error("Unable to resolve endpoint for service", zap.String("service", serviceName), zap.String("region", region), zap.Error(err))
```
is it the right approach to continue the process if `getServiceEndpoint` failed? Will `endpoint` be nil in this case?

liustve replied:
`endpoint` would default to the top level config but looking at it that could be confusing.

for example, if a user configures an invalid service or region and the AWS SDK fails to resolve it to an endpoint, we would end up forwarding data to X-Ray (or whichever service is configured by default) rather than telling the user the endpoint failed to parse.

mxiamxia replied:
is this expected? should it be failed instead of failing back to using x-ray endpoint? Basically, the proxy configure has to be valid to continue the routing. fallback is not an option

liustve replied:
ack. Yes this was initially intended as we wanted the behavior to be fallback with this feature

liustve committed:
fix routing rule naming, fix tests and adjust readme to be more specific

liustve committed:
skip signing for invalid rule configurations

liustve committed:
fix fallback logic for region

liustve committed:
fix fallback logic and tests

liustve committed:
simplify endpoint resolution logic

liustve committed:
fix: remove defaultAWSEndpoint fallback, add role_arn comment, update READMEs

liustve committed:
fix: remove defaultAWSEndpoint fallback, add role_arn comment, update READMEs, export RoutingRule

vastin commented on receiver/awsxrayreceiver/README.md:
This README file should not be changed.

liustve replied:
Removed changes.

vastin commented on internal/aws/proxy/cfg.go:
```diff
@@ -62,6 +62,31 @@ type Config struct {
+}
+
+// RoutingRule defines routing configuration for a specific service.
+type RoutingRule struct {
```
It probably needs to export `RoutingRule` in `extension/awsproxy/config.go` for external use.

liustve replied:
Done.

liustve commented:
Not stale. @sky333999 could you remove the stale tag?

liustve committed:
Merge branch 'aws-cwa-dev' into aws-proxy-multi-server-support

liustve committed:
Merge branch 'aws-cwa-dev' into aws-proxy-multi-server-support

movence commented on internal/aws/proxy/server.go:
```diff
@@ -90,10 +97,47 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
+			apiName := strings.TrimPrefix(req.URL.Path, "/")
+
+			// Skip signing for invalid routing rules
+			if serviceConfig, exists := apiRouteMap[apiName]; exists && serviceConfig == nil {
```
when `serviceConfig` is nil then exits early, req won't have URL or Host set which might cause unexpected behavior.

jefchien replied:
Are we sure we want to skip the routing? Would it make sense to fallback on the behavior from before the routing rules were introduced?

srprash replied:
I think we should reject the requests early to notify users about the misconfiguration. IMO, falling back to some default behavior is confusing if the request is going to fail anyway.

movence commented on internal/aws/proxy/server.go:
```diff
@@ -90,10 +97,47 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
-			req.URL.Scheme = awsURL.Scheme
-			req.URL.Host = awsURL.Host
-			req.Host = awsURL.Host
+			apiName := strings.TrimPrefix(req.URL.Path, "/")
```
`buildRoutingMaps` doesn't trim the leading "/". is this just a safe guard or `apiRouteMap` should trim prefix at build time?

liustve replied:
if I recall I think this is intended, the reason that I trim it here is that routing map will not have the leading "/"

jefchien replied:
Are you sure? `TestBuildRoutingMapsWithLeadingSlash` asserts that
```
assert.Equal(t, "logs", apiMap["/PutLogEvents"].ServiceName)
```
Doesn't that mean that the leading "/" is there? This trimmed `apiName` (`PutLogEvents`) will never match the stored `/PutLogEvents`. Think you need to trim it in both places.

srprash replied:
@jefchien I think you're right. I will make the change.

movence commented on internal/aws/proxy/server.go:
```diff
@@ -161,3 +205,82 @@ func setResolverConfig() func(*endpoints.Options) {
+	}
+
+	for i := range routes {
+		route := &routes[i]
```
this should copy `route` object rather than modifying it directly which also modifies `cfg.AdditionalRoutingRules`

jefchien commented on internal/aws/proxy/server.go:
```diff
@@ -90,10 +97,47 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
+				}
+			}
+
+			targetURL, err := url.Parse(endpoint)
```
nit: Seems like this parsing could be done on start up. Could be part of `buildRoutingMaps` since the endpoints aren't changing for different requests to the same API.

If we don't want to change that, then we could also skip the parsing if `serviceConfig.AWSEndpoint` is empty. We already have the URL for `awsEndPoint` with the `awsURL`, which we're currently using as a fallback.

```go
var endpoint string
...
if serviceConfig := apiRouteMap[apiName]; serviceConfig != nil {
  ...
  if serviceConfig.AWSEndpoint != "" {
    endpoint = serviceConfig.AWSEndpoint
  }
}
targetURL := awsURL
if endpoint != "" {
  url, err := url.Parse(endpoint)
  ...
}
```

jefchien commented on internal/aws/proxy/server_test.go:
```diff
@@ -244,3 +242,592 @@ func (m *mockReadCloser) Read(_ []byte) (n int, err error) {
 }
+
+func TestBuildRoutingMapsEmpty(t *testing.T) {
+	apiMap, signerMap := buildRoutingMaps(nil, "", nil, "", &awsutil.AWSSessionSettings{}, nil)
```
nit: Recommend using `zap.NewNop()` instead of a nil logger. If `buildRoutingMaps` ever tries to log something, a nil logger will panic.

jefchien commented on internal/aws/proxy/server.go:
```diff
@@ -90,10 +97,47 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
+			apiName := strings.TrimPrefix(req.URL.Path, "/")
+
+			// Skip signing for invalid routing rules
+			if serviceConfig, exists := apiRouteMap[apiName]; exists && serviceConfig == nil {
```
nit: Don't need to have a separate apiRouteMap lookup for it being nil and it existing.

```go
serviceConfig, exists := apiRouteMap[apiName]
if exists && serviceConfig == nil {
   ...
}

serviceName := cfg.ServiceName
region := *awsCfg.Region
endpoint := awsEndPoint
reqSigner := signer

if serviceConfig != nil {
   ...
}
```

srprash committed:
fix: reject invalid routing rules with HTTP error and avoid mutating config

Reject requests for paths with invalid routing rules by returning a 502
error from a wrapping handler instead of returning early from the Director
with req.URL/Host unset. Copy route objects in buildRoutingMaps to avoid
modifying cfg.AdditionalRoutingRules in place.

liustve committed:
Merge pull request #1 from srprash/aws-proxy-multi-server-support

fix: reject invalid routing rules and avoid mutating config

srprash committed:
fix: address review comments on routing and endpoint parsing

- Move endpoint URL parsing out of per-request Director; default to
  pre-parsed awsURL and only parse when a routing rule sets a custom
  AWSEndpoint.
- Normalize paths in buildRoutingMaps by trimming leading "/" so lookups
  match the Director's trimmed apiName.
- Change rejection status for invalid routing rules from 502 to 400.
- Extract wrapper handler into proxyHandler struct so tests can still
  access the underlying ReverseProxy for transport injection.
- Use zap.NewNop() instead of nil logger in TestBuildRoutingMapsEmpty.

liustve committed:
Merge pull request #2 from srprash/aws-proxy-multi-server-support-review-fixes

fix: address review comments on routing and endpoint parsing

jefchien commented on internal/aws/proxy/server.go:
```diff
@@ -104,21 +141,44 @@ func NewServer(cfg *Config, logger *zap.Logger) (Server, error) {
+	apiName := strings.TrimPrefix(r.URL.Path, "/")
+	if serviceConfig, exists := h.apiRouteMap[apiName]; exists && serviceConfig == nil {
+		h.logger.Warn("Rejecting request for path with invalid routing rule", zap.String("path", apiName))
+		http.Error(w, "invalid routing configuration for path: "+apiName, http.StatusBadRequest)
```
Missing unit test for this check and 400 response.

mitali-salvi committed:
Merge branch 'aws-cwa-dev' into aws-proxy-multi-server-support
